extensions/amp-analytics/0.1/iframe-transport-intersection-observer.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {MessageType} from '../../../src/3p-frame-messaging';
+import {SubscriptionApi} from '../../../src/iframe-helper';
+import {user} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+
+/**
+ * @visibleForTesting
+ */
+export class IframeTransportIntersectionObserver {
+  /**
+   * Constructor
+   * @param {!Window} win The window element.
+   * @param {!HTMLIFrameElement} frame The cross-domain iframe to send
+   *     messages (results of observations) to.
+   */
+  constructor(win, frame) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!HTMLIFrameElement} */
+    this.frame_ = frame;
+
+    /** @private {!Object} */
+    this.intersectionObserverOptions_ = {
+      threshold: [0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5,
+        0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 1.0],
+    };
+
+    /** @private {!Object<string, !Element>} */
+    this.creativeIdToTarget_ = {};
+
+    /** @private {!Object<string, !IntersectionObserver>} */
+    this.creativeIdToIntersectionObserver_ = {};
+
+    /** @private {!../../../src/iframe-helper.SubscriptionApi} */
+    this.postMessageApi_ = new SubscriptionApi(this.frame_,
+        MessageType.SEND_INTERSECTION_OBSERVER_EVENTS, true, req => {
+          if (!req) {
+            return;
+          }
+          const reqCreativeId = req['creativeId'];
+          if (reqCreativeId &&
+            this.creativeIdToTarget_[reqCreativeId] &&
+            !this.creativeIdToIntersectionObserver_[reqCreativeId]) {
+            this.creativeIdToIntersectionObserver_[reqCreativeId] =
+              new IntersectionObserver(
+                  this.onIntersectionObserverEvent_.bind(this, reqCreativeId),
+                  this.intersectionObserverOptions_);
+            this.creativeIdToIntersectionObserver_[reqCreativeId].observe(
+                this.creativeIdToTarget_[reqCreativeId]);
+          }
+        });
+  }
+
+  /**
+   * Adds a creativeId to the list of ids we are listening to
+   * IntersectionObserver events for.
+   * @param {!string} creativeId The creativeId used to identify <the
```

Comment:
Nit: No need for the `!` on primitive types.

Reply:
Done